lib/explorer/shared.ex

```diff
@@ -270,71 +274,176 @@ defmodule Explorer.Shared do
+
+    infer_struct(map, inner, preferred, strict)
+  end
 
-  defp infer_type(item, type) do
-    if infer_type(item) == type do
+  defp infer_type(item, type, preferred, strict) do
+    if infer_type(item, preferred, strict) == type do
       type
     else
       raise ArgumentError,
             "the value #{inspect(item)} does not match the inferred dtype #{inspect(type)}"
     end
   end
 
-  defp infer_type(%Date{} = _item), do: :date
-  defp infer_type(%Time{} = _item), do: :time
-  defp infer_type(%NaiveDateTime{} = _item), do: {:datetime, :microsecond}
-  defp infer_type(%Explorer.Duration{precision: precision} = _item), do: {:duration, precision}
-  defp infer_type(item) when is_integer(item), do: {:s, 64}
-  defp infer_type(item) when is_float(item) or item in @non_finite, do: {:f, 64}
-  defp infer_type(item) when is_boolean(item), do: :boolean
-  defp infer_type(item) when is_binary(item), do: :string
-  defp infer_type(list) when is_list(list), do: infer_list(list, :null)
-  defp infer_type(%{} = map), do: infer_struct(map, nil)
-  defp infer_type(item), do: raise(ArgumentError, "unsupported datatype: #{inspect(item)}")
+  defp infer_type(%Date{} = _item, _preferred, _strict), do: :date
+  defp infer_type(%Time{} = _item, _preferred, _strict), do: :time
+  defp infer_type(%NaiveDateTime{} = _item, _preferred, _strict), do: {:datetime, :microsecond}
 
-  defp infer_list(list, type) do
-    {:list, Enum.reduce(list, type, &infer_type/2)}
+  defp infer_type(%Explorer.Duration{precision: precision} = _item, _preferred, _strict),
+    do: {:duration, precision}
+
+  defp infer_type(item, _, false) when is_integer(item), do: {:s, 64}
+
+  defp infer_type(item, nil, true) when is_integer(item) do
+    if item > 9_223_372_036_854_775_807 do
+      raise_mismatched_dtype!(item, {:s, 64})
+    else
+      {:s, 64}
+    end
   end
 
-  defp infer_struct(%{} = map, types) do
+  defp infer_type(item, preferred, true) when is_integer(item) do
+    case preferred do
+      {:s, 8} when item > -129 and item < 128 ->
+        preferred
+
+      {:s, 16} when item > -32_769 and item < 32_768 ->
+        preferred
+
+      {:s, 32} when item > -2_147_483_649 and item < 2_147_483_648 ->
+        preferred
+
+      {:s, 64} when item > -9_223_372_036_854_775_809 and item < 9_223_372_036_854_775_808 ->
+        preferred
+
+      {:s, _} ->
+        raise_mismatched_dtype!(item, preferred)
+
+      {:u, _} when item < 0 ->
+        raise_mismatched_dtype!(item, preferred)
+
+      {:u, 8} when item < 256 ->
+        preferred
+
+      {:u, 16} when item < 65_536 ->
+        preferred
+
+      {:u, 32} when item < 4_294_967_296 ->
+        preferred
+
+      {:u, 64} when item < 18_446_744_073_709_551_616 ->
+        preferred
+
+      {:u, _} ->
+        raise_mismatched_dtype!(item, preferred)
+
+      _ ->
+        {:s, 64}
+    end
+  end
+
+  defp infer_type(item, _preferred, _strict) when is_float(item) or item in @non_finite,
+    do: {:f, 64}
```

Comment:
Btw, we also have precision issues with floats, since f32 cannot represent as large numbers as f64. :)

Reply:
But that should probably be another pull request.